File: kb-ui/src/types/ops.ts
/**
 * 运维使用分析类型（GET /api/ops/usage）。
 *
 * 回答的是**使用**问题而不是基础设施问题：有没有人在用、用户问了什么、系统答不上来
 * 哪些、哪个检索范式在真正承载流量。数据源是 serving 写的 serving_query_logs。
 */

/** 窗口内的关键数。分母为 0 时后端给 0 而不是 null——前端直接拿去渲染百分比。 */
export interface OpsUsageSummary {
  queries: number
  no_result: number
  /** 0–1 的小数，不是百分数。 */
  no_result_rate: number
  /** 用 P95 不是平均值：平均值会被一堆快查询稀释，掩掉真正卡住人的那条尾巴。 */
  p95_duration_ms: number
  avg_duration_ms: number
  active_paradigms: number
}

/** 答不上来的问题。**这是整份数据里最有行动价值的一段**——它直接指向该补的知识。 */
export interface OpsNoResultQuery {
  /** 用户输入原文（后端截断到 200 字）。仅 admin 可见。 */
  query_text: string
  count: number
  last_at: string | null
}

export interface OpsTopQuery {
  query_text: string
  count: number
  /** 「问得多又答不上」的优先级最高，所以热门榜也带零结果数。 */
  no_result: number
}

export interface OpsParadigmUsage {
  /** `__legacy__` = 走旧 SearchService、没经范式引擎的流量。 */
  paradigm_id: string
  calls: number
  no_result: number
  p95_duration_ms: number
}

/** 每日检索量。后端已补零，前端直接照数组画。 */
export interface OpsTrendPoint {
  date: string
  queries: number
  no_result: number
}

export interface OpsUsage {
  /**
   * false = serving_query_logs 不存在（serving 从没启动过）。此时其余字段是形状相同的
   * 空壳，不是「用量为 0」——页面要说"尚未产生检索日志"，不能画一屏 0。
   */
  available: boolean
  /** 摘要与各列表的窗口天数。 */
  days: number
  /** 趋势折线的窗口天数，与 days 独立。 */
  trend_days: number
  summary: OpsUsageSummary
  no_result_queries: OpsNoResultQuery[]
  top_queries: OpsTopQuery[]
  paradigms: OpsParadigmUsage[]
  trend: OpsTrendPoint[]
  intents: Record<string, number>
  channels: Record<string, number>
}

File: knowledge_mining/mining/api/routes/ops.py
"""运维使用分析 —— GET /api/ops/usage。

概览页的运维区块（摘要）与 设置→系统状态（明细）共用这一个端点。回答的是**使用**
问题而不是基础设施问题：有没有人在用、用户问了什么、系统答不上来哪些、哪个检索范式
在真正承载流量。

**admin-only（require_admin，现查库）**。两个理由，缺一不可：
- 响应里含 `no_result_queries` / `top_queries` 的**用户输入原文**。那是别人打进检索框的
  东西，不该对普通成员敞开。
- 服务级运维数据对无权处理的人只是噪声。

数据来自 serving 拥有的 `serving_query_logs`（跨服务只读，取舍见 query_log_db 模块
docstring）。表不存在时回 available=False 而不是 500 —— serving 从没启动过的部署里
它本来就不存在。
"""
from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Depends, Query, Request

from knowledge_mining.mining.api.domain_scope import require_domain
from knowledge_mining.mining.infra.query_log_db import QueryLogStats
from knowledge_mining.mining.kb.auth import require_admin

router = APIRouter(prefix="/api/ops", tags=["ops"])

# 摘要与明细列表的窗口（天）。7 天：短到能反映"最近怎么样"，长到能盖住一个工作周的
# 周末低谷——3 天的窗口每逢周一都会显示成断崖。
DEFAULT_WINDOW_DAYS = 7

# 趋势折线固定 30 天，与挖掘趋势对齐，不随 days 变：两条折线并排看时窗口必须一致。
TREND_DAYS = 30

NO_RESULT_LIMIT = 10
TOP_QUERY_LIMIT = 20


def _empty_payload(days: int) -> dict[str, Any]:
    """serving 没产出过日志时的空壳。

    保持与正常响应**完全相同的形状**，让前端只判 available、不必对每个字段判
    undefined —— 少一个分支就少一处 "reading 'queries' of undefined"。
    """
    return {
        "available": False,
        "days": days,
        "trend_days": TREND_DAYS,
        "summary": {
            "queries": 0, "no_result": 0, "no_result_rate": 0.0,
            "p95_duration_ms": 0.0, "avg_duration_ms": 0.0, "active_paradigms": 0,
        },
        "no_result_queries": [],
        "top_queries": [],
        "paradigms": [],
        "trend": [],
        "intents": {},
        "channels": {},
    }


@router.get("/usage")
async def usage_stats(
    request: Request,
    domain: str = Query(..., min_length=1),
    days: int = Query(DEFAULT_WINDOW_DAYS, ge=1, le=90),
    _admin: dict[str, Any] = Depends(require_admin),
) -> dict[str, Any]:
    """检索使用分析。口径 = 该域全部检索流量（不按 KB 收敛——这是域级运维视角）。

    与 /api/kb/stats 的可见集收敛不同：那个是「我的知识库」，这个是「这个域被怎么用」。
    admin 本来就能看全域，再按 KB 收一遍既无意义又会让「有多少流量没走范式」失真。
    """
    resolved = require_domain(domain)
    # 查询日志跟着域走，用域池而不是默认池（与 /api/runs 同一套取池方式）
    pool = await request.app.state.domain_pools.async_pool(resolved)
    stats = QueryLogStats(pool)

    if not await stats.is_available():
        return _empty_payload(days)

    return {
        "available": True,
        "days": days,
        "trend_days": TREND_DAYS,
        "summary": await stats.summary(domain=resolved, days=days),
        "no_result_queries": await stats.no_result_queries(
            domain=resolved, days=days, limit=NO_RESULT_LIMIT,
        ),
        "top_queries": await stats.top_queries(
            domain=resolved, days=days, limit=TOP_QUERY_LIMIT,
        ),
        "paradigms": await stats.paradigm_usage(domain=resolved, days=days),
        "trend": await stats.trend(domain=resolved, days=TREND_DAYS),
        "intents": await stats.breakdown(domain=resolved, days=days, column="intent"),
        "channels": await stats.breakdown(domain=resolved, days=days, column="channel"),
    }
